python/tvm/relax/frontend/onnx_frontend.py

```diff
@@ -0,0 +1,1158 @@
+from tvm.script import tir as T, relax as R
+
+
+def new_var(var_name, shape, dtype="float32"):
+    return testing.nn.Parameter(shape=shape, dtype=dtype, name=var_name)
+
+
+def get_type(elem_type):
+    """Converts onnx integer datatype to numpy datatype"""
+    # If a string was passed instead of a tensor type, it does not need
+    # conversion and can be returned.
+    if isinstance(elem_type, str):
+        return elem_type
+
+    try:
+        from onnx.mapping import TENSOR_TYPE_TO_NP_TYPE
+    except ImportError as e:
+        raise ImportError("Unable to import onnx which is required {}".format(e))
+
+    try:
+        from onnx import TensorProto
+    except ImportError as e:
+        raise ImportError("Unable to import TensorProto from onnx {}".format(e))
+
+    return str(TENSOR_TYPE_TO_NP_TYPE[elem_type])
+
+
+def get_info(info_proto):
+    """Extract the shape from a ValueInfoProto."""
+    shape = []
+    shape_name = []
+    for dim in info_proto.type.tensor_type.shape.dim:
+        name = dim.dim_param
+        value = dim.dim_value
+        if value is None or value == 0:
+            value = tvm.tir.Var("d", "int64")
+            shape_name.append(name)
+        else:
+            shape_name.append(value)
+        shape.append(value)
+
+    name = info_proto.name
+    if info_proto.type.tensor_type.elem_type:
+        dtype = get_type(info_proto.type.tensor_type.elem_type)
+    else:
+        dtype = None
+    return name, shape, dtype, shape_name
+
+
+def get_numpy(tensor_proto):
+    """Grab data in TensorProto and convert to numpy array."""
+    try:
+        from onnx.numpy_helper import to_array
+    except ImportError as e:
+        raise ImportError("Unable to import onnx which is required {}".format(e))
+    return to_array(tensor_proto)
+
+
+class onnx_input(list):
+    """A helper extension to list that returns None for out of bound indices."""
+
+    def __getitem__(self, item):
+        if isinstance(item, slice):
+            if item.stop is None:
+                stop = len(self)
+            else:
+                stop = item.stop
+            indices = list(range(stop)[item])
+            return [self[i] for i in indices]
+        if isinstance(item, int):
+            return list(self)[item] if item < len(self) else None
+        raise TypeError("list indices must be integers or slices, not %s" % type(item).__name__)
+
+
+class OnnxOpConverter(object):
+    """A helper class for holding onnx op converters."""
+
+    @classmethod
+    def get_converter(cls, opset):
+        """Get converter matches given opset.
+
+        Parameters
+        ----------
+        opset: int
+            opset from model.
+
+        Returns
+        -------
+        converter, which should be `_impl_vx`. Number x is the biggest
+            number smaller than or equal to opset belongs to all support versions.
+        """
+        versions = [int(d.replace("_impl_v", "")) for d in dir(cls) if "_impl_v" in d]
```

Comment:
Might there be a simpler way to accomplish the dispatching by opset? E.g., as a property in the converter classes. Encoding it in the method names seems a little unusual. Is split the only operator that has multiple versions in this manner?

Reply:
In principle, additionally, we should test the error case for rejecting an op due to version (we could test this function separately)

Reply:
At the moment -- yes, split is the only operator that has multiple versions. Probably in the future, this will not hold true. 

I am not sure about using a different kind of encoding -- what is nice with this implementation is that each converter does not have to contain a different mapping opset version -> function implementation